waku/waku_sync/protocol.nim

```diff
@@ -0,0 +1,475 @@
+  transferCallBack: Option[TransferCallback] # Callback for message transfer.
+
+  pruneCallBack: Option[PruneCallBack] # Callback with the result of the archive query
+  pruneStart: Timestamp # Last pruning start timestamp
+  pruneOffset: timer.Duration # Offset to prune a bit more than necessary. 
+
+  periodicSyncFut: Future[void]
+  periodicPruneFut: Future[void]
+
+proc storageSize*(self: WakuSync): int =
+  self.storage.len
+
+proc ingessMessage*(self: WakuSync, pubsubTopic: PubsubTopic, msg: WakuMessage) =
+  if msg.ephemeral:
+    return
+
+  let msgHash: WakuMessageHash = computeMessageHash(pubsubTopic, msg)
+
+  trace "inserting message into waku sync storage ",
+    msg_hash = msgHash.to0xHex(), timestamp = msg.timestamp
+
+  if self.storage.insert(msg.timestamp, msgHash).isErr():
+    error "failed to insert message ", msg_hash = msgHash.to0xHex()
+
+proc calculateRange(jitter: Duration, syncRange: Duration = 1.hours): (int64, int64) =
+  var now = getNowInNanosecondTime()
+
+  # Because of message jitter inherent to Relay protocol
+  now -= jitter.nanos
+
+  let range = syncRange.nanos
+
+  let start = now - range
+  let `end` = now
+
+  return (start, `end`)
+
+proc request(
+    self: WakuSync, conn: Connection
+): Future[Result[seq[WakuMessageHash], string]] {.async, gcsafe.} =
+  let (start, `end`) = calculateRange(self.relayJitter)
+
+  let initialized =
+    ?clientInitialize(self.storage, conn, self.maxFrameSize, start, `end`)
+
+  debug "sync session initialized",
+    client = self.peerManager.switch.peerInfo.peerId,
+    server = conn.peerId,
+    frameSize = self.maxFrameSize,
+    timeStart = start,
+    timeEnd = `end`
+
+  var hashes: seq[WakuMessageHash]
+  var reconciled = initialized
+
+  while true:
+    let sent = ?await reconciled.send()
+
+    trace "sync payload sent",
+      client = self.peerManager.switch.peerInfo.peerId,
+      server = conn.peerId,
+      payload = reconciled.payload
+
+    let received = ?await sent.listenBack()
+
+    trace "sync payload received",
+      client = self.peerManager.switch.peerInfo.peerId,
+      server = conn.peerId,
+      payload = received.payload
+
+    reconciled = (?received.clientReconcile(hashes)).valueOr:
+      let completed = error # Result[Reconciled, Completed]
+
+      ?await completed.clientTerminate()
+
+      debug "sync session ended gracefully",
+        client = self.peerManager.switch.peerInfo.peerId, server = conn.peerId
+
+      return ok(hashes)
+
+proc sync*(
+    self: WakuSync, peerInfo: Option[RemotePeerInfo] = none(RemotePeerInfo)
+): Future[Result[(seq[WakuMessageHash], RemotePeerInfo), string]] {.async, gcsafe.} =
+  let peer =
+    if peerInfo.isSome():
+      peerInfo.get()
+    else:
+      let peer: RemotePeerInfo = self.peerManager.selectPeer(WakuSyncCodec).valueOr:
+        return err("No suitable peer found for sync")
+
+      peer
+
+  let connOpt = await self.peerManager.dialPeer(peer, WakuSyncCodec)
```

Comment:
We may dial the same peer if we only have only one but unlikely if we have more. AFAIK we select a random sync peer since we don't specify which one we want.

Reply:
@gabrielmer AFAIK no other protocol close their connection either so it seams fine? :thinking: Maybe they all should.